highlighters/main/main-highlighter.zsh

```diff
@@ -677,8 +678,21 @@ _zsh_highlight_main_highlighter_highlight_list()
      # try-always construct
      style=reserved-word # de facto a reserved word, although not de jure
      next_word=':start:'
+   elif [[ $this_word == :function: ]]; then
+     if [[ $arg == '()' ]]; then
+       style=reserved-word
+       next_word=:start:
+     else
+       style=function-definition
+       next_word=:function:
+     fi
    elif ! (( in_redirection)) && [[ $this_word == *':start:'* ]]; then # $arg is the command word
-     if (( ${+precommand_options[$arg]} )) && _zsh_highlight_main__is_runnable $arg; then
+     if [[ $res != reserved && ( $args[1] == '()' ||
+           # TODO: Function names can be absurd; this handles the more common cases without invoking Cthulhu.
+           ( $zsyh_user_options[multifuncdef] == on && $args[(r)*[^[:alnum:]_-]*] == '()' ) ) ]]; then
```

Comment:
I wonder if we could do something along the line of "after `function`, everything that isn't a token (`&&`, `>`, etc) or the reserved words `{` or `()` is a function name?  Haven't looked at Src/parse.c of this.

Reply:
`function` already has that in effect. This branch only effects words before `()`.

Reply:
Sorry, that's what I meant. "If one of the following words is `()`, and it's in the same simple command as the current word, then all intervening words [other than redirections] are function names" or something like this.  Basically, we know that "everything from the start of the command to the `()` token is function names", with some exceptions (`noglob`, redirections, …), so I wonder if a blacklist might be better than a whitelist.